src/core/PluginPinConnector.cpp

```diff
@@ -0,0 +1,517 @@
+		if (numRouted > 0)
+		{
+			// Normalize output
+			for (f_cnt_t frame = 0; frame < frames; ++frame)
+			{
+				outPtr[frame][0] = buffer[frame] / numRouted;
+			}
+		}
+
+		// Right SampleFrame channel second
+		mixInputs(outChannel + 1, numRouted, buffer);
+		if (numRouted > 0)
+		{
+			// Normalize output
+			for (f_cnt_t frame = 0; frame < frames; ++frame)
+			{
+				outPtr[frame][1] = buffer[frame] / numRouted;
+			}
+		}
+
+		// NOTE: When numRouted == 0, nothing needs to be written to `inOut` for
+		// the audio bypass, since it already contains the LMMS core input audio.
+	}
+}
+
+void PluginPinConnector::saveSettings(QDomDocument& doc, QDomElement& elem)
+{
+	auto pins = doc.createElement(nodeName());
+	elem.appendChild(pins);
+
+	if (m_trackChannelsUsed != s_totalTrackChannels)
+	{
+		pins.setAttribute("tc_used", m_trackChannelsUsed);
+	}
+
+	pins.setAttribute("num_in", in().channelCount);
+	pins.setAttribute("num_out", out().channelCount);
+
+	auto pinsIn = doc.createElement("in_matrix");
+	pins.appendChild(pinsIn);
+	saveSettings(m_in, doc, pinsIn);
+
+	auto pinsOut = doc.createElement("out_matrix");
+	pins.appendChild(pinsOut);
+	saveSettings(m_out, doc, pinsOut);
+}
+
+void PluginPinConnector::loadSettings(const QDomElement& elem)
+{
+	const auto pins = elem.firstChildElement(nodeName());
+	if (pins.isNull()) { return; }
+
+	// Until full routing support is added, track channel count should always be 2
+	m_trackChannelsUsed = pins.attribute("tc_used", QString::number(s_totalTrackChannels)).toUInt();
+	assert(m_trackChannelsUsed == 2);
+
+	// TODO: Assert port counts are what was expected?
+	const auto pluginInCount = pins.attribute("num_in", "0").toInt();
+	const auto pluginOutCount = pins.attribute("num_out", "0").toInt();
+	setPluginChannelCounts(pluginInCount, pluginOutCount);
+
+	loadSettings(pins.firstChildElement("in_matrix"), m_in);
+	loadSettings(pins.firstChildElement("out_matrix"), m_out);
+}
+
+void PluginPinConnector::saveSettings(const Matrix& matrix, QDomDocument& doc, QDomElement& elem)
+{
+	// Only saves connections that are actually used, otherwise could bloat project file
+	for (std::size_t trackChannel = 0; trackChannel < matrix.pins.size(); ++trackChannel)
+	{
+		auto& trackChannels = matrix.pins[trackChannel];
+		for (std::size_t pluginChannel = 0; pluginChannel < trackChannels.size(); ++pluginChannel)
+		{
+			if (trackChannels[pluginChannel]->value() || trackChannels[pluginChannel]->isAutomatedOrControlled())
+			{
+				trackChannels[pluginChannel]->saveSettings(doc, elem,
+					QString{"c%1_%2"}.arg(trackChannel + 1).arg(pluginChannel + 1));
+			}
+		}
+	}
+}
+
+void PluginPinConnector::loadSettings(const QDomElement& elem, Matrix& matrix)
+{
+	const auto trackChannelCount = static_cast<int>(matrix.pins.size());
+	const auto pluginChannelCount = trackChannelCount > 0 ? static_cast<int>(matrix.pins[0].size()) : 0;
+	assert(pluginChannelCount == matrix.channelCount);
```

Comment:
I changed this a little. I think I can use `channelCount` here instead of messing with `pins[0].size()`, so I'm doing that. But I kept the `pins[0].size()` check in an assert just so that I can know that assumption holds.